## Data Acquisition

This process will attempt to get a variety of posts on weibo in terms of topic, time, and individual/organizations/corporates/government.

There is a lack of archives for past Weibo posts, which means it will need to be obtained through API calls. However, from what I know, the only possible method to retreive data of Weibo posts is through their API calls which requires a query to be made (blank querying unfortunately does not work).

In order to generate topics that traces back as much as possible, and as broad as possible, I tried multiple ways, but I ended up with using Selenium to simulate a browser to webscrape the website: https://www.weibotop.cn/2.0/ which contains a small archive of the top 50 "hot topics" in the past. Detailed to the minute, this archive stores the topic name, its popularity, and rank out of 50. Considering that this was only for a more elaborate/broad dataset of posts, the popularity and changes in rank does not matter. So I chose to sample the hot topics every week on Monday.

The website lagged a lot, the changes reflected by the change of dates often does not show up until a few minutes later. So the topic acquisition was often repetitive and length. Moreover, the earliest data it has is in 2019 October, which only provides us with at most 6 years of hot topics to query from.

In [ ]:
pip install requests pyquery selenium webdriver-manager

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 36.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 492.9/492.9 kB 26.4 MB/s eta 0:00:00


In [ ]:
!apt-get update
!apt-get install -y google-chrome-stable
!pip install selenium webdriver-manager

Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Get:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,372 kB]
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,673 kB]
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:11 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:12 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,235 kB]
Hit:13 https://ppa.launchpadconten

Trying Inspurer Github's method.

This unfortunately did not work becase accessing the website required logging in.

In [ ]:
import requests
from lxml import etree

# Weibo Hot Search URL
WEIBO_HOT_URL = "https://s.weibo.com/top/summary"

# User-Agent to mimic a real browser request
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:66.0) Gecko/20100101 Firefox/66.0',
    'Cookie': 'SUB = _2A25K1UC3DeRhGeBN6FER9SbFyDqIHXVpq9x_rDV6PUJbktANLRTlkW1NRCE-4IHhCmLgD1lfqX5KQGh-sSL4Iezw'
}

def get_hot_topics():
    """Scrapes Weibo's hot search page and returns a list of trending topics."""
    response = requests.get(WEIBO_HOT_URL, headers=headers)

    if response.status_code == 200:
        html_content = response.text
        print(html_content[:1000])

        html = etree.HTML(response.text)
        # Extract trending topics
        hot_topics = html.xpath("//td[@class='td-02']/a/text()")
        return hot_topics[:10] #just to sample
    else:
        print("Failed to fetch trending topics.")
        return []

if __name__ == "__main__":
    trending_topics = get_hot_topics()
    print("Trending Weibo Topics:", trending_topics)
    #despite there being cookies, this still doesnt work which suggests that alternative methods may be required

In [ ]:
response = requests.get(WEIBO_HOT_URL, headers=headers)
print(response.text)
# Page source still shows that clicking into this page leads to the Sina Weibo Visitor's page

<!DOCTYPE html>
<html>
<head>
    <meta http-equiv="Content-type" content="text/html; charset=gb2312"/>
    <title>Sina Visitor System</title>
</head>
<body>
<span id="message"></span>
<script type="text/javascript" src="/js/visitor/mini_original.js?v=20161116"></script>
<script type="text/javascript">
    window.use_fp = "1" == "1"; // ÊÇ·ñ²É¼¯É豸ָÎÆ¡£
    var url = url || {};
    (function () {
        this.l = function (u, c) {
            try {
                var s = document.createElement("script");
                s.type = "text/javascript";
                s[document.all ? "onreadystatechange" : "onload"] = function () {

                    if (document.all && this.readyState != "loaded" && this.readyState != "complete") {
                        return
                    }
                    this[document.all ? "onreadystatechange" : "onload"] = null;
                    this.parentNode.removeChild(this);
                    if (c) {
                        c()
          

Here I was trying to grab from the Weibo's Hot Topic page directly through simulating a browser using Selenium. However, I realized that it only has the most recent data and refreshes every minute, so it did not fit my purpose.

In [ ]:
!apt update
!apt install -y wget curl unzip
!wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
!dpkg -i google-chrome-stable_current_amd64.deb || true
!apt --fix-broken install -y
!rm google-chrome-stable_current_amd64.deb
!google-chrome --version
#to have a path to google chrome on google colab for selenium to simulate

Hit:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:3 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease
Hit:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
29 packages can be upgraded. Run 'apt list --upgradable' to see them.
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InR

In [ ]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By

options = webdriver.ChromeOptions()
options.binary_location = "/usr/bin/google-chrome-stable"
options.add_argument("--headless")
options.add_argument("--disable-gpu")
options.add_argument("--no-sandbox")
options.add_argument("--disable-dev-shm-usage")

# Initialize WebDriver
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)

driver.get("https://s.weibo.com/top/summary")
driver.implicitly_wait(10)  # Wait for JavaScript to load

hot_topics = [element.text for element in driver.find_elements(By.XPATH, "//td[@class='td-02']/a")]

driver.quit()

print("Trending Weibo Topics:", hot_topics)
#this works, so the next step is to complete this with a for loop added with sleep time to prevent bot detections

Trending Weibo Topics: ['习近平的两会足迹', '海底捞小便事件10倍现金补偿', 'DeepSeek眼中不会被AI替代的职业', '2025奋进中国', '海底捞补偿金额超千万元', '韩国低于性同意年龄均视为强奸', '铜川一医院被曝偷换新生儿', '章昊FILAFUSION潮流品牌大使', '海底捞已起诉两名小便男子', '杨铭宇黄焖鸡隔夜鸡肉优先出售', '桃李等畅销面包频现发霉异物问题', '杨铭宇黄焖鸡道歉', '韩媒再曝金秀贤金赛纶亲密照', '周深用英语向中国日报读者打招呼', '刚baby baby几天就塌了', '眼泪女王', '高位截瘫患者用意念控制机械臂写字', 'D社证实金秀贤金赛纶恋爱关系', '海底捞', '不懂金秀贤做错什么了的韩男们', '你要是嗑CP', '被指倒扣工资公司注册资本达7亿', '刘雯走秀鞋子掉了', '孙中山逝世100周年', '百万网红千惠自曝患焦虑症', '海底捞小便事件', '章若楠素颜镜头', '黄明昊 你们好甜啊', '林一角色号发了虞书欣未公开图', 'papi看到胡彦斌的手串眼里都在放光', '买辆车签3个合同3款车型', '夸克居然可以个性化定制面经', '郑钦文2比0科斯秋克', '近视手术是一种奢侈美容', '希望台湾同胞一道继承孙中山先生遗愿', '张凌赫种植豆芽', '杨铭宇董事长曾称食品安全是生命线', '林一新剧追星成功武大靖', '塞尔吉尼奥入籍', '女生三年攒160个纸袋卖了180元', '商务部等部门约谈沃尔玛', '王楚钦3比0阿鲁纳', 'AD钙奶这个味儿谁研究出来的', '王一博胡兵 胡萝卜', '嘘国王在冬眠 武大靖', '买不起黄金的年轻人盯上了债基', '年轻人开始流行各种自带了', '直播间买的百万金条不发货后续来了', '董明珠说为格力省了十几亿广告费', '少爷和我杀回来了', '嘘国王在冬眠 胡文煊', '杨铭宇黄焖鸡正联系涉事门店关店整改']


In [ ]:
import time
import random
import pandas as pd
from datetime import datetime
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By

options = webdriver.ChromeOptions()
options.binary_location = "/usr/bin/google-chrome-stable"
options.add_argument("--headless")
options.add_argument("--disable-gpu")
options.add_argument("--no-sandbox")
options.add_argument("--disable-dev-shm-usage")

driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)

all_hot_topics = set()
iteration = 0

try:
    while True:
        iteration += 1
        print(f"\n Scraping Weibo Hot Topics - Iteration {iteration} at {datetime.now()}")

        try:
            driver.get("https://s.weibo.com/top/summary")
            driver.implicitly_wait(10)

            hot_topics = [element.text for element in driver.find_elements(By.XPATH, "//td[@class='td-02']/a")]
            all_hot_topics.update(hot_topics)

            # Save periodically every 10 iterations
            if iteration % 20 == 0:
                df = pd.DataFrame({"Trending Topics": list(all_hot_topics)})
                df.to_csv("topics.csv", index=False, encoding="utf-8-sig")
                print(f"Saved {len(all_hot_topics)} unique topics to {csv_filename}")

        except Exception as e:
            print(f"Error encountered: {e}")

        # Wait randomly between 5-20 seconds before the next request
        sleep_time = random.randint(5, 20)
        print(f"Sleeping for {sleep_time} seconds...")
        time.sleep(sleep_time)

except KeyboardInterrupt:
    print("\nStopping the scraper manually...")

    # Final save before exit
    df = pd.DataFrame({"Trending Topics": list(all_hot_topics)})
    df.to_csv("topics.csv", index=False, encoding="utf-8-sig")

    driver.quit()

    #this doesnt work because the hot topics are only the most recent ones where they refresh every 1 minute, to get hsitorical data we would require other datasets.


 Scraping Weibo Hot Topics - Iteration 1 at 2025-03-12 07:35:39.619147
Sleeping for 14 seconds...

 Scraping Weibo Hot Topics - Iteration 2 at 2025-03-12 07:35:57.775825
Sleeping for 7 seconds...

 Scraping Weibo Hot Topics - Iteration 3 at 2025-03-12 07:36:06.065631
Sleeping for 17 seconds...

 Scraping Weibo Hot Topics - Iteration 4 at 2025-03-12 07:36:23.970976
Sleeping for 13 seconds...

 Scraping Weibo Hot Topics - Iteration 5 at 2025-03-12 07:36:37.844608
Sleeping for 7 seconds...

 Scraping Weibo Hot Topics - Iteration 6 at 2025-03-12 07:36:45.715630
Sleeping for 10 seconds...

 Scraping Weibo Hot Topics - Iteration 7 at 2025-03-12 07:36:58.315551
Sleeping for 12 seconds...

 Scraping Weibo Hot Topics - Iteration 8 at 2025-03-12 07:37:11.192362
Sleeping for 12 seconds...

 Scraping Weibo Hot Topics - Iteration 9 at 2025-03-12 07:37:24.050344
Sleeping for 20 seconds...

 Scraping Weibo Hot Topics - Iteration 10 at 2025-03-12 07:37:44.949689
Sleeping for 18 seconds...

 Scraping 


Stopping the scraper manually...


This is the method that I ended up using, where I was able to scrape the archived Hot Topics for querying specific Weibo posts.

In [ ]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import pandas as pd
import time
from webdriver_manager.chrome import ChromeDriverManager
from selenium.common.exceptions import StaleElementReferenceException, TimeoutException, ElementClickInterceptedException

# Set up Selenium WebDriver
options = webdriver.ChromeOptions()
options.add_argument("--headless")  # Run in headless mode (no GUI)
options.add_argument("--no-sandbox")
options.add_argument("--disable-dev-shm-usage")

driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)

# Open Weibo Top Archive website
url = "https://www.weibotop.cn/2.0/"
driver.get(url)

# Wait until the date input field is available
try:
    WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, "selectDateTime")))
except TimeoutException:
    print("Failed to locate the date input field. Exiting.")
    driver.quit()
    exit()

# Generate weekly dates from 2019-10-28 (first Monday) to 2025-03-10
start_date = pd.to_datetime("2019-10-28")
end_date = pd.to_datetime("2025-03-10")
date_range = pd.date_range(start=start_date, end=end_date, freq='W-MON')  # Every Monday

data = []
save_interval = 10  # Save progress every 10 iterations
csv_filename = "weibo_hot_search_weekly.csv"

for i, date in enumerate(date_range):
    date_str = date.strftime("%m%d%Y")  # MMDDYYYY format

    try:
        # Locate the date input field in every iteration to avoid stale references
        date_input = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, "selectDateTime")))

        # Try clicking the element, handling intercepts
        try:
            date_input.click()
        except ElementClickInterceptedException:
            driver.execute_script("arguments[0].click();", date_input)

        time.sleep(1)

        # Enter date manually
        date_input.send_keys(date_str[:2])  # Enter Month (MM)
        time.sleep(0.5)
        date_input.send_keys(date_str[2:4])  # Enter Day (DD)
        time.sleep(0.5)
        date_input.send_keys(date_str[4:])  # Enter Year (YYYY)
        time.sleep(0.5)

        # Click outside to trigger the search
        driver.find_element(By.TAG_NAME, "body").click()
        time.sleep(15)

        # Extract trending topics
        topics = driver.find_elements(By.XPATH, "//h5[@class='mb-1']")
        for topic in topics:
            data.append([date.strftime("%Y-%m-%d"), topic.text])

    except StaleElementReferenceException:
        print(f"⚠️ StaleElementReferenceException encountered on {date.strftime('%Y-%m-%d')}, retrying...")
        continue  # Skip to the next date
    except TimeoutException:
        print(f"⚠️ TimeoutException: No topics found for {date.strftime('%Y-%m-%d')}, skipping...")
        continue

    # Periodic saving every 10 iterations
    if (i + 1) % save_interval == 0:
        df = pd.DataFrame(data, columns=["Date", "Topic"])
        df.to_csv(csv_filename, mode='a', header=not bool(i), index=False)  # Append mode
        print(f"💾 Progress saved at {date.strftime('%Y-%m-%d')} ({i + 1}/{len(date_range)})")
        data = []  # Clear in-memory data to free memory

# Final save for remaining data
if data:
    df = pd.DataFrame(data, columns=["Date", "Topic"])
    df.to_csv(csv_filename, mode='a', header=False, index=False)
    print("✅ Final save completed.")

# Close WebDriver
driver.quit()
print(f"🎉 Data saved to {csv_filename}")

💾 Progress saved at 2019-12-30 (10/281)
💾 Progress saved at 2020-03-09 (20/281)
💾 Progress saved at 2020-05-18 (30/281)
💾 Progress saved at 2020-07-27 (40/281)
💾 Progress saved at 2020-10-05 (50/281)
💾 Progress saved at 2020-12-14 (60/281)
💾 Progress saved at 2021-02-22 (70/281)
💾 Progress saved at 2021-05-03 (80/281)
💾 Progress saved at 2021-07-12 (90/281)
💾 Progress saved at 2021-09-20 (100/281)
💾 Progress saved at 2021-11-29 (110/281)
💾 Progress saved at 2022-02-07 (120/281)
💾 Progress saved at 2022-04-18 (130/281)


KeyboardInterrupt: 

In [ ]:
data

[['2025-03-10', '告诉你8月更新的新功能'],
 ['2025-03-10', '高源批评江一燕'],
 ['2025-03-10', '未成年人犯罪父母有啥责'],
 ['2025-03-10', '易烊千玺吻戏'],
 ['2025-03-10', '马思纯平头'],
 ['2025-03-10', '活埋婴儿爷爷被刑拘'],
 ['2025-03-10', '装39人集装箱到港口时已密封'],
 ['2025-03-10', '未成年人刑责年龄该降低吗'],
 ['2025-03-10', '中国赴黎巴嫩维和部队开展义诊活动'],
 ['2025-03-10', '代拍乱象'],
 ['2025-03-10', 'Luna发文悼念雪莉'],
 ['2025-03-10', '公交车司机过站该不该停车'],
 ['2025-03-10', '浙大通报69人因呕吐腹泻就诊'],
 ['2025-03-10', '军运会上一个人的比赛'],
 ['2025-03-10', '40名大学生因旷课太多被退学'],
 ['2025-03-10', '英国民众为集装箱39名死者守夜'],
 ['2025-03-10', '姐弟恋最甜蜜的时刻'],
 ['2025-03-10', '罗红'],
 ['2025-03-10', '巴黎奥运会logo被指太性感'],
 ['2025-03-10', '少年的你影评'],
 ['2025-03-10', '易烊千玺演技'],
 ['2025-03-10', '男女生用毛巾的区别'],
 ['2025-03-10', '文化水平赶不上小学生'],
 ['2025-03-10', '孟晚舟为任正非过生日'],
 ['2025-03-10', '晒丑指甲大赛'],
 ['2025-03-10', '黄婷婷回复黑粉'],
 ['2025-03-10', '鹿晗新剧杀青'],
 ['2025-03-10', '林心如公开示爱霍建华'],
 ['2025-03-10', '英国首相寻求提前大选'],
 ['2025-03-10', '你大爷大妈还是你大爷大妈'],
 ['2025-03-10', '郑爽 荧光绿无袖上衣'],
 ['2025-03-10', '得罪造型师的影视剧'],
 ['2025-03-10', '陈情令之生

In [ ]:
data

[['2019-10-28', '告诉你8月更新的新功能'],
 ['2019-10-28', '高源批评江一燕'],
 ['2019-10-28', '未成年人犯罪父母有啥责'],
 ['2019-10-28', '易烊千玺吻戏'],
 ['2019-10-28', '马思纯平头'],
 ['2019-10-28', '活埋婴儿爷爷被刑拘'],
 ['2019-10-28', '装39人集装箱到港口时已密封'],
 ['2019-10-28', '未成年人刑责年龄该降低吗'],
 ['2019-10-28', '中国赴黎巴嫩维和部队开展义诊活动'],
 ['2019-10-28', '代拍乱象'],
 ['2019-10-28', 'Luna发文悼念雪莉'],
 ['2019-10-28', '公交车司机过站该不该停车'],
 ['2019-10-28', '浙大通报69人因呕吐腹泻就诊'],
 ['2019-10-28', '军运会上一个人的比赛'],
 ['2019-10-28', '40名大学生因旷课太多被退学'],
 ['2019-10-28', '英国民众为集装箱39名死者守夜'],
 ['2019-10-28', '姐弟恋最甜蜜的时刻'],
 ['2019-10-28', '罗红'],
 ['2019-10-28', '巴黎奥运会logo被指太性感'],
 ['2019-10-28', '少年的你影评'],
 ['2019-10-28', '易烊千玺演技'],
 ['2019-10-28', '男女生用毛巾的区别'],
 ['2019-10-28', '文化水平赶不上小学生'],
 ['2019-10-28', '孟晚舟为任正非过生日'],
 ['2019-10-28', '晒丑指甲大赛'],
 ['2019-10-28', '黄婷婷回复黑粉'],
 ['2019-10-28', '鹿晗新剧杀青'],
 ['2019-10-28', '林心如公开示爱霍建华'],
 ['2019-10-28', '英国首相寻求提前大选'],
 ['2019-10-28', '你大爷大妈还是你大爷大妈'],
 ['2019-10-28', '郑爽 荧光绿无袖上衣'],
 ['2019-10-28', '得罪造型师的影视剧'],
 ['2019-10-28', '陈情令之生

In [ ]:
searches = pd.read_csv("/content/weibo_hot_search_weekly.csv", names=["Date", "Topic"])

In [ ]:
 topics = searches["Topic"].unique()
 topic_list= topics.tolist()

Acquiring Weibo posts from these topics

In [ ]:
import json
import requests
import time
import os

# Base URL template
url_template = (
    'https://m.weibo.cn/api/container/getIndex'
    '?type=wb'
    '&queryVal={query_val}'
    '&containerid=100103type=2%26q%3D{query_val}'
    '&page={page_num}'
)

def fetch_data(query_val, page_id):
    # Format the URL with the query parameters
    url = url_template.format(query_val=query_val, page_num=page_id)
    print(f"Fetching URL: {url}")

    # Add headers to mimic a browser
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3',
        'Referer': 'https://m.weibo.cn/',
        'X-Requested-With': 'XMLHttpRequest'
    }

    try:
        resp = requests.get(url, headers=headers)

        if resp.status_code != 200:
            print(f"Error fetching data: HTTP {resp.status_code}")
            return [], False  # Return empty list and False to indicate no more data

        data = resp.json()
        # Check if the response contains 'cards' and it's not empty
        cards = data.get('data', {}).get('cards', [])
        if not cards:
            print("No more cards found.")
            return [], False

        print(f"Fetched {len(cards)} cards from page {page_id}")
        return cards, True  # Return the cards and True to continue fetching
    except Exception as e:
        print(f"Error fetching or parsing data: {e}")
        return [], False

def extract_mblogs(cards):
    mblogs = []
    for card in cards:
        try:
            mblog = card.get('mblog', {})
            blog = {
                'mid': mblog.get('id'),
                'text': mblog.get('text'),
                'userid': str(mblog.get('user', {}).get('id', '')),
                'username': mblog.get('user', {}).get('screen_name', ''),
                'created_at': mblog.get('created_at'),
                'reposts_count': mblog.get('reposts_count', 0),
                'comments_count': mblog.get('comments_count', 0),
                'attitudes_count': mblog.get('attitudes_count', 0)
            }
            mblogs.append(blog)
        except KeyError as e:
            print(f"Missing key in card: {e}")
    return mblogs

def save_backup(mblogs, filename="weibo_data.json"):
    """ Appends new data to an existing JSON file. """
    if os.path.exists(filename):
        with open(filename, 'r', encoding='utf-8') as f:
            try:
                existing_data = json.load(f)
            except json.JSONDecodeError:
                existing_data = []
    else:
        existing_data = []

    existing_data.extend(mblogs)  # Add new data

    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(existing_data, f, ensure_ascii=False, indent=4)

    print(f"Saved backup: {len(existing_data)} total posts in {filename}")

def fetch_all_pages(query_val):
    mblogs = []
    page_id = 1  # Start from page 1

    while True:  # Loop until there are no more pages
        cards, has_more_data = fetch_data(query_val, page_id)
        if not has_more_data:
            break  # Stop if no more data is returned

        # Extract mblogs from the cards
        page_mblogs = extract_mblogs(cards)
        mblogs.extend(page_mblogs)

        print(f"Total posts collected so far for '{query_val}': {len(mblogs)}")
        page_id += 1  # Move to the next page

        # Save a backup every 10 pages
        if page_id % 10 == 0:
            save_backup(mblogs)
            mblogs = []  # Clear the list after saving

        # Add a delay to avoid rate-limiting
        time.sleep(2)

    # Final save if there are remaining mblogs
    if mblogs:
        save_backup(mblogs)

def fetch_multiple_queries(query_list):
    for query_val in query_list:
        print(f"\nFetching data for query: '{query_val}'")
        fetch_all_pages(query_val)

In [ ]:
# List of query values
query_list = topic_list

# Fetch data for all queries
fetch_multiple_queries(query_list)

Streaming output truncated to the last 5000 lines.
Fetched 10 cards from page 79
Total posts collected so far for '罗红': 91
Saved backup: 17334 total posts in weibo_data.json
Fetching URL: https://m.weibo.cn/api/container/getIndex?type=wb&queryVal=罗红&containerid=100103type=2%26q%3D罗红&page=80
Fetched 10 cards from page 80
Total posts collected so far for '罗红': 10
Fetching URL: https://m.weibo.cn/api/container/getIndex?type=wb&queryVal=罗红&containerid=100103type=2%26q%3D罗红&page=81
Fetched 10 cards from page 81
Total posts collected so far for '罗红': 20
Fetching URL: https://m.weibo.cn/api/container/getIndex?type=wb&queryVal=罗红&containerid=100103type=2%26q%3D罗红&page=82
Fetched 8 cards from page 82
Total posts collected so far for '罗红': 28
Fetching URL: https://m.weibo.cn/api/container/getIndex?type=wb&queryVal=罗红&containerid=100103type=2%26q%3D罗红&page=83
Fetched 10 cards from page 83
Total posts collected so far for '罗红': 38
Fetching URL: https://m.weibo.cn/api/container/getIndex?type=wb&que

## Data Preprocessing

With the scraped data, I have to clean the text section to remove the extra hashtag content and the hyperlinks, as well as texts that were not posted by the poster so that it is easier to identify and predict misogyny

In [ ]:
import pandas as pd

In [ ]:
import json

with open("weibo_data.json", "r", encoding="utf-8") as f:
    try:
        data = json.load(f)
    except json.JSONDecodeError:
        print("JSON is corrupted or incomplete. Trying to fix...")
        f.seek(0)  # Go back to the start of the file
        lines = f.readlines()[:-1]  # Remove the last line (potentially incomplete)

        # Try loading again after removing the last line
        data = json.loads("".join(lines))

df = pd.DataFrame(data)

In [ ]:
df.head()
#the text contains a lot of hyperlinks and emojis which takes up tokens and also makes misogyny identification harder

/usr/local/lib/python3.11/dist-packages/google/colab/_dataframe_summarizer.py:88: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  cast_date_col = pd.to_datetime(column, errors="coerce")


mid                                               text  \
0              None                                               None   
1  5143674116374623  <a  href="https://m.weibo.cn/search?containeri...   
2  5143674086231653  <a  href="https://m.weibo.cn/search?containeri...   
3  5143673977703294  <a  href="https://m.weibo.cn/p/index?extparam=...   
4  5143673839815410  【<a  href="https://m.weibo.cn/search?container...   

       userid  username                      created_at  reposts_count  \
0                                                  None              0   
1  7388961691     小龙猫猫_  Thu Mar 13 09:55:36 +0800 2025              0   
2  3383588712       花酱团  Thu Mar 13 09:55:30 +0800 2025              0   
3  1212878871     来了啊来福  Thu Mar 13 09:55:04 +0800 2025              0   
4  5251226673  绍虞城事通网绍兴  Thu Mar 13 09:54:31 +0800 2025              0   

   comments_count  attitudes_count  
0               0                0  
1               0                0  
2               0                0  
3               0                0  
4               0                0

In [ ]:
df.shape

(127497, 8)

In [ ]:
df["text"][6] #there is a lot of information that isn't required

'<a  href="https://m.weibo.cn/search?containerid=231522type%3D1%26t%3D10%26q%3D%23%E9%A1%BE%E8%8C%9C%E8%8C%9C%E6%8A%96%E9%9F%B3%E8%B4%A6%E5%8F%B7%E8%A2%AB%E6%B0%B8%E4%B9%85%E5%B0%81%E7%A6%81%23&extparam=%23%E9%A1%BE%E8%8C%9C%E8%8C%9C%E6%8A%96%E9%9F%B3%E8%B4%A6%E5%8F%B7%E8%A2%AB%E6%B0%B8%E4%B9%85%E5%B0%81%E7%A6%81%23" data-hide=""><span class="surl-text">#顾茜茜抖音账号被永久封禁#</span></a>怎么回事？就暴了一下收入被封，是暴露了这个行业多好赚，被同行搞？ '

In [ ]:
import re

In [ ]:
def clean_text(text):
    if not text:
        return ""
    # Remove hyperlinks and anchor tags
    text = re.sub(r"http\S+|www\S+|<a.*?>|</a>", "", text)
    # Remove all span classes (surl-text, url-icon, and others)
    text = re.sub(r"<span.*?>.*?</span>", "", text)
    # Remove hashtags (content inside ##)
    text = re.sub(r"#.*?#", "", text)
    # Remove any remaining HTML tags
    text = re.sub(r"<.*?>", "", text)
    # Remove emojis and special characters
    text = re.sub(r"[^\w\s\u4e00-\u9fff]", "", text)
    return text.strip()

# Apply the function to the text column
df["cleaned_text"] = df["text"].apply(clean_text)

# Display cleaned text
df[["text", "cleaned_text"]].head()

text  \
0                                               None   
1  <a  href="https://m.weibo.cn/search?containeri...   
2  <a  href="https://m.weibo.cn/search?containeri...   
3  <a  href="https://m.weibo.cn/p/index?extparam=...   
4  【<a  href="https://m.weibo.cn/search?container...   

                                        cleaned_text  
0                                                     
1                                活该 一直都不是很喜欢 有点太妹的感觉  
2                           599SK2神仙水230ml老款日期到26年4月  
3            百万粉丝网红顾茜茜被永久封禁 曾自曝躺一天啥也没干赚30万元老话说得好树大招风  
4  近日此前自曝日入超30万元引发争议拥有近500万粉丝的主播顾茜茜抖音账号被正式永久封禁企查查...

In [ ]:
df['cleaned_text'][6]

'怎么回事就暴了一下收入被封是暴露了这个行业多好赚被同行搞'

taking a sample of weibo dataset for training

In [ ]:
#saving this for future reference:
df.to_csv("weibo_original_cleaned_df.csv", index=False, encoding="utf-8")

In [ ]:
sample_size = 500
df_sample = df.sample(n=sample_size, random_state=42)
df_sample["misogyny_label"] = 0 #creating the column for misogyny label

# Save sample for manual labeling
df_sample.to_csv("weibo_sample_for_labeling.csv", index=False, encoding="utf-8")

Simultaneously, because this data covers (or tries to cover) as many topics as possible, it may not provide enough consideration for what is considered misogynistic and what is not. So I wanted to include my previous datasets as well for training.

In [ ]:
feminist = pd.read_csv("/content/feministdiscourse.csv")

In [ ]:
feminist.head()
#for this dataset, i have previously ran sentiment analysis, collected similar weibo information, and manually labeled them. -1 is feminist, 0 neutral, and 1 is misogynistic.

mid                                               text  \
0  4744160006113240  最后一小时。拼手速时间。劵马上🈚️。一小时后涨20块钱。。。王小卤的虎皮鸡爪，多好吃不说了哈...   
1  4867011886449270                       极端偏激女权主义分子特征，善良的男性们，你们要警惕啊！    
2  4757606730039870  人们对于不认同的男性发言，带上的标签是“大男子主义”或者是对其个人的剖析，对于不认同的女性发...   
3  5102628826251560  媎妹感觉你智商为0//<a href='/n/Spidpire'>@Spidpire</a>...   
4  5109772649763750  最近女硕士失踪13年被“收留”大孩子12岁的新闻很火。我根本不想讨论，因为没有意义了：<br...   

       userid         username                 created_at  reposts_count  \
0  1833569404          辛德瑞拉小卖部  2022-03-06 23:09:10+08:00              0   
1  6444840645         好时代见证记录者  2023-02-08 23:18:41+08:00              6   
2  7514394135  Seok_Cheon上班摸鱼版  2022-04-13 01:41:39+08:00             93   
3  1837223697           玟咚Yusi  2024-11-20 03:36:17+08:00              0   
4  5522145332          小小瑞_171  2024-12-09 20:43:17+08:00              0   

   comments_count  attitudes_count keyword  sentiment_score snownlp_sentiment  \
0              36              118     媎妹          0.549855           neutral   
1              22               49   极端女权          0.975390          positive   
2             141             1826   田园女拳          0.294325          negative   
3               0                0     媎妹          0.023232          negative   
4               0                6   女性权利          0.999536          positive   

   misogyny_label  
0               0  
1               1  
2              -1  
3               1  
4              -1

In [ ]:
feminist_cleaned = feminist.drop(columns=["keyword", "sentiment_score", "snownlp_sentiment"])

In [ ]:
feminist_cleaned["cleaned_text"] = feminist_cleaned["text"].apply(clean_text) #remove unecessary symbols and hyperlinks

feminist_cleaned["misogyny_label"] = feminist_cleaned["misogyny_label"].replace(-1, 0) #i'm not going to identify what is and isn't feminist for this project

feminist_cleaned.head()

mid                                               text  \
0  4744160006113240  最后一小时。拼手速时间。劵马上🈚️。一小时后涨20块钱。。。王小卤的虎皮鸡爪，多好吃不说了哈...   
1  4867011886449270                       极端偏激女权主义分子特征，善良的男性们，你们要警惕啊！    
2  4757606730039870  人们对于不认同的男性发言，带上的标签是“大男子主义”或者是对其个人的剖析，对于不认同的女性发...   
3  5102628826251560  媎妹感觉你智商为0//<a href='/n/Spidpire'>@Spidpire</a>...   
4  5109772649763750  最近女硕士失踪13年被“收留”大孩子12岁的新闻很火。我根本不想讨论，因为没有意义了：<br...   

       userid         username                 created_at  reposts_count  \
0  1833569404          辛德瑞拉小卖部  2022-03-06 23:09:10+08:00              0   
1  6444840645         好时代见证记录者  2023-02-08 23:18:41+08:00              6   
2  7514394135  Seok_Cheon上班摸鱼版  2022-04-13 01:41:39+08:00             93   
3  1837223697           玟咚Yusi  2024-11-20 03:36:17+08:00              0   
4  5522145332          小小瑞_171  2024-12-09 20:43:17+08:00              0   

   comments_count  attitudes_count  misogyny_label  \
0              36              118               0   
1              22               49               1   
2             141             1826               0   
3               0                0               1   
4               0                6               0   

                                        cleaned_text  
0  最后一小时拼手速时间劵马上一小时后涨20块钱王小卤的虎皮鸡爪多好吃不说了哈哈哈反正就是一小时...  
1                           极端偏激女权主义分子特征善良的男性们你们要警惕啊  
2  人们对于不认同的男性发言带上的标签是大男子主义或者是对其个人的剖析对于不认同的女性发言带上的...  
3  媎妹感觉你智商为0Spidpire说白了就是没钱给工资呗一年12kw从去年开始压合同价格压到...  
4  最近女硕士失踪13年被收留大孩子12岁的新闻很火我根本不想讨论因为没有意义了韩女我的胯下 不...

In [ ]:
feminist_cleaned["cleaned_text"][2]

'人们对于不认同的男性发言带上的标签是大男子主义或者是对其个人的剖析对于不认同的女性发言带上的标签是境外势力极端女权田园女拳拳师有没有官方声讨过那些以侮辱女性为己任的男人没有因为他们不是极端男权而是概括为没有素质或者没有文化'

In [ ]:
feminist_cleaned.shape #500 rows

(500, 10)

In [ ]:
feminist_cleaned.to_csv("feminist_sample_for_labeling.csv", index=False, encoding="utf-8") #i went through the sample again to see if this is what i wanted

merging the dataset and training the macBERT model

In [ ]:
feminist_labeled = pd.read_csv("/content/labeled_feminist.csv")
feminist_labeled["mid"] = feminist_labeled["mid"].astype(str).str.split('00%').str[0]

In [ ]:
weibo_labeled = pd.read_csv("/content/labeled_weibo.csv")
weibo_labeled[['mid', 'userid']] = weibo_labeled[['mid', 'userid']].astype(str)

In [ ]:
feminist_labeled.sample(10)

mid                                               text  \
175  5109452143332700              女性要发声，要努力站到权利中心，为解放女性不懈奋斗，宁愿痛苦，不要麻木。    
165  5103911826100250  回复<a href='/n/无敌的泰兰特又倒下了'>@无敌的泰兰特又倒下了</a>:女法官协...   
371  5106033402320410  我觉得一般情况下的直女，因为对同性没什么兴趣，所以对同性的标准应该是很低的，倾向于跟同性建立...   
372  5023106533689790  点进去看了两段半 都没看到几爱豆出现 点进主页果然是男的<br />男的就是不能讨论女权 别...   
46   5110484992527950  性骚扰！每位女网红绕不过去的话题…… <a  href="https://video.wei...   
359  5107240143160120  38岁大龄剩女彩礼要80万，你们觉得怎么样？ <a  href="https://video...   
21   5110719362371880                            眼睛袅袅了…世界上有女孩子就是最美好的一件事情   
469  5104475731919010  哈哈哈哈哈哈哈哈哈哈看到的最有趣的一句话，好了，可以安心的睡觉了，哈哈哈哈哈我只能说锐评<s...   
273  5103122515233690  <a  href="https://m.weibo.cn/search?containeri...   
153  4981815750428540  女权真TM不是人，人家消防员夫妻相互恩爱，她们在评论区诅咒人家分开。虽然这是消防员，不是军人...   

     misogyny_label    username                 created_at  reposts_count  \
175               0     哪里来的毛毛雨  2024-12-08 23:29:42+08:00              0   
165               0    轻武器高工王_辉  2024-11-23 16:34:28+08:00              6   
371               0      盐不浪老板娘  2024-11-29 13:04:51+08:00              3   
372               0  Morble有鱼就烤  2024-04-14 17:03:05+08:00              1   
46                0        苒苒学姐  2024-12-11 19:53:53+08:00              0   
359               1        趣味表哥  2024-12-02 21:00:01+08:00              8   
21                0  jiangmanil  2024-12-12 11:25:11+08:00              0   
469               1       个人日记簿  2024-11-25 05:55:14+08:00              0   
273               1        宝儿贰爷  2024-11-21 12:18:02+08:00              0   
153               1       一个坏土豆  2023-12-22 18:28:15+08:00             17   

     comments_count  attitudes_count  \
175               1                0   
165               4               15   
371               7              152   
372               2               24   
46                7                5   
359               7               13   
21                0                0   
469               0                0   
273               0                4   
153              71              476   

                                          cleaned_text  
175                    女性要发声要努力站到权利中心为解放女性不懈奋斗宁愿痛苦不要麻木  
165  回复无敌的泰兰特又倒下了女法官协会这是公开地搞司法中的女拳组织了吧无敌的泰兰特又倒下了回复a...  
371  我觉得一般情况下的直女因为对同性没什么兴趣所以对同性的标准应该是很低的倾向于跟同性建立一种平...  
372  点进去看了两段半 都没看到几爱豆出现 点进主页果然是男的男的就是不能讨论女权 别说什么剪了再...  
46                                     性骚扰每位女网红绕不过去的话题  
359                               38岁大龄剩女彩礼要80万你们觉得怎么样  
21                              眼睛袅袅了世界上有女孩子就是最美好的一件事情  
469  哈哈哈哈哈哈哈哈哈哈看到的最有趣的一句话好了可以安心的睡觉了哈哈哈哈哈我只能说锐评 运动酱r...  
273                                      动了谁的蛋糕谁的媎妹小编急  
153  女权真TM不是人人家消防员夫妻相互恩爱她们在评论区诅咒人家分开虽然这是消防员不是军人但是照着...

In [ ]:
weibo_labeled.drop(columns=['userid'], inplace=True)

In [ ]:
weibo_labeled.shape

(500, 9)

In [ ]:
mlready = pd.concat([feminist_labeled, weibo_labeled])

In [ ]:
mlready.head()

mid                                               text  \
0  4744160006113240  最后一小时。拼手速时间。劵马上🈚️。一小时后涨20块钱。。。王小卤的虎皮鸡爪，多好吃不说了哈...   
1  4867011886449270                       极端偏激女权主义分子特征，善良的男性们，你们要警惕啊！    
2  4757606730039870  人们对于不认同的男性发言，带上的标签是“大男子主义”或者是对其个人的剖析，对于不认同的女性发...   
3  5102628826251560  媎妹感觉你智商为0//<a href='/n/Spidpire'>@Spidpire</a>...   
4  5109772649763750  最近女硕士失踪13年被“收留”大孩子12岁的新闻很火。我根本不想讨论，因为没有意义了：<br...   

   misogyny_label         username                 created_at  reposts_count  \
0               0          辛德瑞拉小卖部  2022-03-06 23:09:10+08:00              0   
1               1         好时代见证记录者  2023-02-08 23:18:41+08:00              6   
2               0  Seok_Cheon上班摸鱼版  2022-04-13 01:41:39+08:00             93   
3               1           玟咚Yusi  2024-11-20 03:36:17+08:00              0   
4               0          小小瑞_171  2024-12-09 20:43:17+08:00              0   

   comments_count  attitudes_count  \
0              36              118   
1              22               49   
2             141             1826   
3               0                0   
4               0                6   

                                        cleaned_text  
0  最后一小时拼手速时间劵马上一小时后涨20块钱王小卤的虎皮鸡爪多好吃不说了哈哈哈反正就是一小时...  
1                           极端偏激女权主义分子特征善良的男性们你们要警惕啊  
2  人们对于不认同的男性发言带上的标签是大男子主义或者是对其个人的剖析对于不认同的女性发言带上的...  
3  媎妹感觉你智商为0Spidpire说白了就是没钱给工资呗一年12kw从去年开始压合同价格压到...  
4  最近女硕士失踪13年被收留大孩子12岁的新闻很火我根本不想讨论因为没有意义了韩女我的胯下 不...

From these attempts at fine-tuning the model, there are some conclusions that could be made that points to ideas for the final project.

**In terms of improvements for this model**

One way to go about it is continuing to tune the hyperparameters, but it may not be significantly improving the finetuning of this model. Based on the previous evaluation metrics, the epoches seems optimal between 3-5. With the above attempts, raising the learning rate has made the training and validation loss higher, which may suggest that it isn't helpful pertaining to the currently available data & training set. Thus, increasing batches also may not be helpful considering the small training set, as it increases the possibility of overfitting.


**Real-Life Significance of the fine-tuning** *- was it necessary to finetune this model?*

The fine-tuning increased its accuracy by 0.01%,which is not a big improvement from a simplistic approach to classify every post as non-misogynistic. This imbalanced dataset makes it difficult to evaluate the actual abilities of this fine-tuned language model in identifying misogyny. Moreover, perhaps the limited training sample sizes made it unfit to be used for MacBERT. As a large-scale transformer model, it usually needs a larger training sample to effectively learn discriminatory features.

The improvement itself was rather insignificant considering that there is likely only 2 posts that are misogynistic out of 500. The accuracy, precision, and recall is likely to fluctuate largely and is unfair to evaluate its performance with this dataset.


**Issues pertaining the data**

While fine-tuning may be possible and beneficial, I argue that given this dataset, it is unlikely that accuracy could be raised higher. This Weibo dataset was pulled through identifying the top 50 "hot topics" （热搜）and with those topics as queries for the Weibo API. Because these are topics that are commonly discussed among Weibo users, it limits the covered topics to be news or entertainment related. Many of the samples posts that I manually labeled had discussions of specific Chinese-pop stars and celebrities, government-affiliated organizations, current news, and product advertisements that tags itself with hot topics. This suggests that this technique of data acquisition would only pull commonly accepted topics, flushing out more personal or niche discussions. Hence why this dataset contrasts with what was pulled for me previous project (in the feministdiscourse.csv) which were much more focused on feminism or misogynistic language. With the queries revolved around more public-oriented posts, it is natural that the language is more tame with very implicit biases; such as referring to women in their fields with "women+job title", or cursing with female-organ related terms (although conventional, but still implicitly biased). Considering the nature of these posts, the number of possible misogynistic posts, and the limited training sample, I argue that the problem lies more with the data than it does with the hyperparameters of the language model. There is not a lot of room for both mistakes or improvements of prediction for the language model, as seen as the dummy accuracy itself is as high as 98%.


This project reminds me that for me to get a better dataset, I may need search for other ways to get information. The first method is to conduct further data mining. However, the current dataset with 120,000 data took around 15 hours. Additionally, the archive of hot topics were limited from 2019 October to present, which does not give depth in terms of time.

Another alternative would be looking into news archives or Chinese Twitter/Reddit posts which may an outlook in other aspects that is different from mainland Chinese feminism/language-usage.


**Alternatives/Outlooks for final project**

An alternative is to switch a model, either through using a smaller pre-trained language model or through LLM APIs. A smaller model could improve the performance of classifying misogyny when given a small training set, like the one I created for this objective. There are two ways to use LLM for improvements. 1) Using LLM as a method to label for the training set: by giving a prompt and giving examples or terms, the LLM may be able to contextualize and help identifying a more subtle misogyny that a face-front language model through tokenizing like BERT may not be able to do. 2) Using LLM directly for classification and predicting, which could be a faster way to go about it.

I would also propose that the data acquisition method be changed so both the training set and the full dataset could be larger and covers both personal and official posts.
